File: Session 6 - Open Addressing With Double Hashing/Uploaded files/OpenAddressingHashTable.cpp
#include <iostream>
#include <string>
#include <vector>

class OpenAddressHashTable {
private:
    int m;
    static const std::string DELETED;
    std::vector<std::string> T;

    int stringToIntegerConverter(const std::string& key) const {
        int intValue = 0;
        for (char ch : key) {
            intValue += static_cast<int>(ch);
        }
        return intValue;
    }

    int hash(const std::string& key, int i) const {
        int value = stringToIntegerConverter(key);
        int h1 = value % m;
        int h2 = value % 107;
        return (h1 + i * h2) % m;
    }

public:
    OpenAddressHashTable(int m) : m(m), T(m, "") {}

    int insert(const std::string& key) {
        if (key.empty() || key == DELETED) {
            throw std::invalid_argument("Key cannot be empty or DELETED");
        }

        for (int i = 0; i < m; i++) {
            int index = hash(key, i);
            if (T[index].empty() || T[index] == DELETED) {
                T[index] = key;
                return i;
            }
        }
        return -1; // Table is full
    }

    int search(const std::string& key) const {
        for (int i = 0; i < m; i++) {
            int index = hash(key, i);
            if (T[index].empty()) {
                return -1; // Key not found
            }
            if (T[index] == key) {
                return index;
            }
        }
        return -1; // Key not found
    }

    bool deleteKey(int index) {
        if (index < 0 || index >= m) {
            return false;
        } else {
            T[index] = DELETED;
            return true;
        }
    }

    void printHash(const std::string& message) const {
        std::cout << "**************\n";
        std::cout << message << std::endl;
        for (const auto& item : T) {
            std::cout << item << std::endl;
        }
    }
};

// Define the DELETED constant
const std::string OpenAddressHashTable::DELETED = std::string();

int main() {
    OpenAddressHashTable openAddressHashTable(10);

    int res;

    // 1
    res = openAddressHashTable.insert("Algorithm");
    openAddressHashTable.printHash("Inserted string: Algorithm");
    std::cout << "Inserted at i = " << res << std::endl;

    // 2
    res = openAddressHashTable.insert("Data Structure");
    openAddressHashTable.printHash("Inserted string: Data Structure");
    std::cout << "Inserted at i = " << res << std::endl;

    // 3
    res = openAddressHashTable.insert("Hello");
    openAddressHashTable.printHash("Inserted string: Hello");
    std::cout << "Inserted at i = " << res << std::endl;

    // 4
    res = openAddressHashTable.insert("mhtiroglA");
    openAddressHashTable.printHash("Inserted string: mhtiroglA");
    std::cout << "Inserted at i = " << res << std::endl;

    // 5
    int index = openAddressHashTable.search("Algorithm");
    bool result = openAddressHashTable.deleteKey(index);
    openAddressHashTable.printHash("Deleted string: Algorithm");
    std::cout << "Result is " << std::boolalpha << result << std::endl;

    // 6
    std::cout << "UNSUCCESSFUL DELETION BECAUSE THIS KEY DOES NOT EXIST IN OUR HASH TABLE" << std::endl;
    index = openAddressHashTable.search("No");
    result = openAddressHashTable.deleteKey(index);
    openAddressHashTable.printHash("Deleted string: No");

    // 7
    std::cout << "UNSUCCESSFUL SEARCH BECAUSE THIS KEY DOES NOT EXIST IN OUR HASH TABLE" << std::endl;
    res = openAddressHashTable.search("Yes");
    openAddressHashTable.printHash("Searched string: Yes");
    std::cout << "Search result is at " << res << std::endl;

    // 8
    res = openAddressHashTable.search("Data Structure");
    openAddressHashTable.printHash("Searched string: Data Structure");
    std::cout << "Search result is at " << res << std::endl;

    // 9
    res = openAddressHashTable.insert("olleH");
    openAddressHashTable.printHash("Inserted: olleH");
    std::cout << "Inserted at i = " << res << std::endl;

    return 0;
}

File: Session 6 - Open Addressing With Double Hashing/Uploaded files/OpenAddressingHashTable.py
class OpenAddressHashTable:
    DELETED = object()

    def __init__(self, m):
        self.m = m
        self.T = [None] * m

    def string_to_integer_converter(self, key):
        int_value = 0
        for char in key:
            int_value += ord(char)
        return int_value

    def hash(self, key, i):
        value = self.string_to_integer_converter(key)
        h1 = value % self.m
        h2 = value % 107
        return (h1 + i * h2) % self.m

    def insert(self, key):
        if key is None or key is self.DELETED:
            raise ValueError("Key cannot be null")

        for i in range(self.m):
            index = self.hash(key, i)
            if self.T[index] is None or self.T[index] is self.DELETED:
                self.T[index] = key
                return i

        return -1

    def search(self, key):
        for i in range(self.m):
            index = self.hash(key, i)
            if self.T[index] is None:
                return -1
            if self.T[index] == key:
                return index
        return -1

    def delete(self, index):
        if index < 0 or index >= self.m:
            return False
        else:
            self.T[index] = self.DELETED
            return True

    def print_hash(self, message):
        print("**************")
        print(message)
        for item in self.T:
            print(item)

if __name__ == "__main__":
    open_address_hash_table = OpenAddressHashTable(10)

    # 1
    res = open_address_hash_table.insert("Algorithm")
    open_address_hash_table.print_hash("Inserted string : Algorithm")
    print("Inserted at i =", res)

    # 2
    res = open_address_hash_table.insert("Data Structure")
    open_address_hash_table.print_hash("Inserted string : Data Structure")
    print("Inserted at i =", res)

    # 3
    res = open_address_hash_table.insert("Hello")
    open_address_hash_table.print_hash("Inserted string : Hello")
    print("Inserted at i =", res)

    # 4
    res = open_address_hash_table.insert("mhtiroglA")
    open_address_hash_table.print_hash("Inserted string : mhtiroglA")
    print("Inserted at i =", res)

    # 5
    index = open_address_hash_table.search("Algorithm")
    result = open_address_hash_table.delete(index)
    open_address_hash_table.print_hash("Deleted string : Algorithm")
    print("Result is", result)

    # 6
    print("UNSUCCESSFUL DELETION BECAUSE THIS KEY DOES NOT EXIST IN OUR HASH TABLE")
    index = open_address_hash_table.search("No")
    result = open_address_hash_table.delete(index)
    open_address_hash_table.print_hash("Deleted string : No")

    # 7
    print("UNSUCCESSFUL SEARCH BECAUSE THIS KEY DOES NOT EXIST IN OUR HASH TABLE")
    res = open_address_hash_table.search("Yes")
    open_address_hash_table.print_hash("Searched string : Yes")
    print("Search result is at", res)

    # 8
    res = open_address_hash_table.search("Data Structure")
    open_address_hash_table.print_hash("Searched string : Data Structure")
    print("Search result is at", res)

    # 9
    res = open_address_hash_table.insert("olleH")
    open_address_hash_table.print_hash("Inserted : olleH")
    print("Inserted at i =", res)
